packages/vite/src/node/plugins/asset.ts

```diff
@@ -381,63 +389,97 @@ async function fileToBuiltUrl(
   const cache = assetCache.get(config)!
   const cached = cache.get(id)
   if (cached) {
-    return cached
+    return cached.url
   }
 
   const file = cleanUrl(id)
   const content = await fsp.readFile(file)
-
   let url: string
-  if (
-    config.build.lib ||
-    (!file.endsWith('.svg') &&
-      !file.endsWith('.html') &&
-      content.length < Number(config.build.assetsInlineLimit))
-  ) {
-    const mimeType = mrmime.lookup(file) ?? 'application/octet-stream'
-    // base64 inlined as a string
-    url = `data:${mimeType};base64,${content.toString('base64')}`
-  } else {
-    // emit as asset
-    // rollup supports `import.meta.ROLLUP_FILE_URL_*`, but it generates code
-    // that uses runtime url sniffing and it can be verbose when targeting
-    // non-module format. It also fails to cascade the asset content change
-    // into the chunk's hash, so we have to do our own content hashing here.
-    // https://bundlers.tooling.report/hashing/asset-cascade/
-    // https://github.com/rollup/rollup/issues/3415
-    const map = assetHashToFilenameMap.get(config)!
-    const contentHash = getHash(content)
-    const { search, hash } = parseUrl(id)
-    const postfix = (search || '') + (hash || '')
-
-    const fileName = assetFileNamesToFileName(
-      resolveAssetFileNames(config),
-      file,
-      contentHash,
-      content
-    )
-    if (!map.has(contentHash)) {
-      map.set(contentHash, fileName)
-    }
-    const emittedSet = emittedHashMap.get(config)!
-    if (!emittedSet.has(contentHash)) {
-      const name = normalizePath(path.relative(config.root, file))
-      pluginContext.emitFile({
-        name,
-        fileName,
-        type: 'asset',
-        source: content
-      })
-      emittedSet.add(contentHash)
+  let size: number
+
+  /*
+  lib should always inlined
+  svg should never be inlined (unless lib)
+  */
+  if (config.build.lib) {
+    url = fileToInlinedAsset(file, content)
+    size = 0
+  } else if (file.endsWith('.svg') === false) {
+    const inlinedURL = fileToInlinedAsset(file, content)
```

Comment:
@yyx990803 Fair enough, but being accurate with size is inherently important as many small add up.
However I got the same feedback internally and the current design is to just pass an object with various methods

So instead of passing three arguments, it's one argument with computed methods; 
```ts 
{
  name: string
  size(): number
  bufferSize(): number
  totalSize(): number
}
```

Let the user decide if it's wasteful or not, at least Vite won't be doing it now unless opt-in, and even then you can choose to do it for e.g just `assets/*`

Will that work for you?

Reply:
I think the object is a good idea, but I don't think we should expose `size()` here. It is a clever trick but we will be allowing users to easily reach for inefficient code. For most use cases, it is ok to use the file size and we should promote that.
I think we should have the object but with a plain `size` prop and not `totalSize`.

Reply:
@patak-dev The problem is bufferSize doesn't match the **REAL** size and that is the concern this PR is trying to solve. 

We have a bunch of assets, some should be inlined and some should not.
File name matters (explicit inlining) 
Paths matter (glob inlining) 
Size matters (It's too big, opt out for this one)
TotalSize (We're reaching our limit, stop inlining) 

Now, all of these could inherently be done by the user in the callback, but it would cumbersome.

for instance, totalSize is very easy to do outside, just `totalSize += realSize` outside the callback.
But expecting the user to know and do

1. That the current size given to you is a lie
2. Getting the real size at runtime

That being said, I've already written this for myself, so it's not a big deal. 

Yeah I agree having an object there makes more sense for any further information needed to be passed down.

Your call
👍🏼 

That being said, I need help with a few things: 

What's the consensus on https://github.com/vitejs/vite/pull/2909 - should SVG be under the inline assets umbrella without base64?

I might need some help resolving the conflict.

Reply:
As you say, if someone needs the `totalSize`, it can be computed externally. I don't think it is common enough to justify having it in core.

And the same goes for the real size, we can be more clear in the docs and add a rationale about why the API is using the file size.

About #2909, I think it should but that PR should be merged first, no? We can already move with this PR before that IIUC.

Let me know if you try to resolve the conflict and you get blocked and I'll check it out 👍🏼